test/sm4_internal_test.c

```diff
@@ -75,12 +75,164 @@ static int test_sm4_ecb(void)
 
     return 1;
 }
+
+static int test_sm4_cipher(const EVP_CIPHER *cipher,
+                           const unsigned char *key,
+                           const unsigned char *iv,
+                           const unsigned char *input,
+                           const unsigned char *expected,
+                           unsigned int len)
+{
+    EVP_CIPHER_CTX *ctx = NULL;
+    unsigned char ctext[1024];
+    unsigned char ptext[1024];
+    int outlen, tmplen;
+    int ret = 0;
+
+    if (!TEST_ptr(cipher)
+            || !TEST_ptr(ctx = EVP_CIPHER_CTX_new())
+            || !TEST_true(EVP_EncryptInit_ex(ctx, cipher, NULL, key, iv))
+            || !TEST_true(EVP_EncryptUpdate(ctx, ctext, &outlen, input, len))
+            || !TEST_true(EVP_EncryptFinal_ex(ctx, ctext + outlen, &tmplen))
+            || !TEST_mem_eq(ctext, len, expected, len))
+        goto err;
+
+    outlen += tmplen;
+    if (!TEST_true(EVP_DecryptInit_ex(ctx, cipher, NULL, key, iv))
+            || !TEST_true(EVP_DecryptUpdate(ctx, ptext, &outlen, ctext, outlen))
```

Comment:
Why don't use `len` instead of `outlen` directly?

Reply:
From the test results, the outlen and len of CBC mode are not equal, and the direct use of len will also cause the decryption to fail. It seems that the lengths of the ciphertext and plaintext in CBC mode are not equal.